connman/src/service.c

```diff
@@ -1553,8 +1556,26 @@ static int nameserver_add_all(struct connman_service *service,
 				service->nameservers[i]);
 			i++;
 		}
+	/*
+	 * If there are no nameservers and there are no fallback nameservers
+	 * defined add the gateway as the DNS server, which in some tethering
+	 * configurations or because of an error might not be sent to clients.
+	 */
+	} else if (!fallback_dns) {
+		int index = __connman_service_get_index(service);
+		const char *dns = __connman_ipconfig_get_gateway_from_index(
+								index, type);
+
+		DBG("No DNS servers or fallback DNS servers set."
+						"Use gw %s as DNS", dns);
+		__connman_service_nameserver_append(service, dns, false);
+
+		nameserver_add(service, type, dns);
```

Comment:
Are both of these really needed? Based on quick look __connman_service_nameserver_append does call nameserver_add but with CONNMAN_IPCONFIG_TYPE_ALL not type, so is the forcing of type needed also?

Reply:
It seems to be that latter can be dropped. The `type` is just to avoid adding the nameserver for a different IP family, or to avoid a check with early bail-out in `nameserver_available()`. Anyways, the IP family of the nameserver is checked there as well, so `nameserver_add()` is not needed here.